Improve performance of system.query_metric_log

pamarcos opened the pull request:
Several performance and accuracy improvements:

- Use a separate mutex for each query to reduce the critical region for queries_mutex which is used on every single query.
- Add TSA annotations to avoid mistakes in the future
- Use `LockGuard` since it contains TSA annotations as opposed to `std::unique_lock`
- Call `QueryMetricLog::finishQuery` ASAP because it takes [up to 2s ](https://github.com/ClickHouse/ClickHouse/issues/71244#issuecomment-2464603210) since the query finishes until it's called

Close #71244

#### Improvement

Even though we should take benchmarking with a grain of salt, I've attempted to collect organic results rather than microbenchmarking.

I've run 10K requests with 1K concurrent connections with hey. The query is a 2.5s sleep where collection is done every 123ms.

```bash
hey -n 10000 -c 1000 -m GET 'http://localhost:8123/?query=select+sleep(2.5)&query_metric_log_interval=123'
```

Then, I've used [ctrack](https://github.com/Compaile/ctrack) since it provides very nice metrics and it's very easy to integrate with an existing codebase. The results show quite an improvement.

##### collectMetric

before

```
+--------------------+------------+------+----------+---------+---------+--------+---------+
|      filename      |  function  | line | time acc |   sd    |   cv    | calls  | threads |
+--------------------+------------+------+----------+---------+---------+--------+---------+
| QueryMetricLog.cpp | operator() |  123 | 296.41 s | 6.17 ms | 4.16414 | 200137 |     512 |
+--------------------+------------+------+----------+---------+---------+--------+---------+
+-----------+-----------+-----------+---------+------------+---------+---------+----------+----------+-----------+
|     fastest[0-1]%     |                          center[1-99]%                          |   slowest[99-100]%   |
+-----------+-----------+-----------+---------+------------+---------+---------+----------+----------+-----------+
|    min    |   mean    |    min    |  mean   |    med     | time a  | time ae |   max    |   mean   |    max    |
+-----------+-----------+-----------+---------+------------+---------+---------+----------+----------+-----------+
| 25.33 mcs | 41.33 mcs | 47.48 mcs | 1.03 ms | 116.35 mcs | 14.85 s |  4.31 s | 30.97 ms | 46.64 ms | 125.47 ms |
+-----------+-----------+-----------+---------+------------+---------+---------+----------+----------+-----------+
```

after

```
+--------------------+---------------+------+----------+---------+--------+--------+---------+
|      filename      |   function    | line | time acc |   sd    |   cv   | calls  | threads |
+--------------------+---------------+------+----------+---------+--------+--------+---------+
| QueryMetricLog.cpp | collectMetric |  119 |  66.14 s | 1.43 ms | 4.3145 | 200005 |     512 |
+--------------------+---------------+------+----------+---------+--------+--------+---------+
+-----------+-----------+-----------+------------+------------+---------+---------+---------+----------+----------+
|     fastest[0-1]%     |                           center[1-99]%                           |  slowest[99-100]%   |
+-----------+-----------+-----------+------------+------------+---------+---------+---------+----------+----------+
|    min    |   mean    |    min    |    mean    |    med     | time a  | time ae |   max   |   mean   |   max    |
+-----------+-----------+-----------+------------+------------+---------+---------+---------+----------+----------+
| 26.44 mcs | 45.38 mcs | 51.07 mcs | 216.80 mcs | 127.50 mcs | 15.51 s |  5.77 s | 5.98 ms | 11.78 ms | 66.45 ms |
+-----------+-----------+-----------+------------+------------+---------+---------+---------+----------+----------+
```

##### createLogMetricElement

This is called within `createMetric`

before 

```
+--------------------+------------------------+------+----------+---------+---------+--------+---------+
|      filename      |        function        | line | time acc |   sd    |   cv    | calls  | threads |
+--------------------+------------------------+------+----------+---------+---------+--------+---------+
| QueryMetricLog.cpp | createLogMetricElement |  194 | 283.59 s | 6.02 ms | 4.46342 | 210137 |    1512 |
+--------------------+------------------------+------+----------+---------+---------+--------+---------+
+-----------+-----------+-----------+------------+-----------+---------+---------+----------+----------+-----------+
|     fastest[0-1]%     |                           center[1-99]%                           |   slowest[99-100]%   |
+-----------+-----------+-----------+------------+-----------+---------+---------+----------+----------+-----------+
|    min    |   mean    |    min    |    mean    |    med    | time a  | time ae |   max    |   mean   |    max    |
+-----------+-----------+-----------+------------+-----------+---------+---------+----------+----------+-----------+
| 16.17 mcs | 25.00 mcs | 27.44 mcs | 909.62 mcs | 48.61 mcs | 10.67 s | 10.67 s | 30.16 ms | 45.79 ms | 125.43 ms |
+-----------+-----------+-----------+------------+-----------+---------+---------+----------+----------+-----------+
```

after

```
+--------------------+------------------------+------+----------+---------+---------+--------+---------+
|      filename      |        function        | line | time acc |   sd    |   cv    | calls  | threads |
+--------------------+------------------------+------+----------+---------+---------+--------+---------+
| QueryMetricLog.cpp | createLogMetricElement |  248 |  56.07 s | 1.49 ms | 5.58397 | 210003 |    1512 |
+--------------------+------------------------+------+----------+---------+---------+--------+---------+
+-----------+-----------+-----------+------------+-----------+---------+---------+---------+----------+----------+
|     fastest[0-1]%     |                          center[1-99]%                           |  slowest[99-100]%   |
+-----------+-----------+-----------+------------+-----------+---------+---------+---------+----------+----------+
|    min    |   mean    |    min    |    mean    |    med    | time a  | time ae |   max   |   mean   |   max    |
+-----------+-----------+-----------+------------+-----------+---------+---------+---------+----------+----------+
| 18.43 mcs | 26.02 mcs | 27.94 mcs | 146.14 mcs | 46.11 mcs | 10.03 s | 10.03 s | 6.83 ms | 12.35 ms | 66.40 ms |
+-----------+-----------+-----------+------------+-----------+---------+---------+---------+----------+----------+
```

##### startQuery

before

```
+--------------------+------------+------+----------+----------+---------+-------+---------+
|      filename      |  function  | line | time acc |    sd    |   cv    | calls | threads |
+--------------------+------------+------+----------+----------+---------+-------+---------+
| QueryMetricLog.cpp | startQuery |  115 |  86.86 s | 13.65 ms | 1.57196 | 10000 |    1000 |
+--------------------+------------+------+----------+----------+---------+-------+---------+
+----------+----------+----------+---------+------------+--------+---------+----------+----------+-----------+
|    fastest[0-1]%    |                         center[1-99]%                         |   slowest[99-100]%   |
+----------+----------+----------+---------+------------+--------+---------+----------+----------+-----------+
|   min    |   mean   |   min    |  mean   |    med     | time a | time ae |   max    |   mean   |    max    |
+----------+----------+----------+---------+------------+--------+---------+----------+----------+-----------+
| 3.50 mcs | 4.62 mcs | 5.09 mcs | 8.17 ms | 123.52 mcs | 2.26 s |  2.26 s | 55.04 ms | 68.00 ms | 126.46 ms |
+----------+----------+----------+---------+------------+--------+---------+----------+----------+-----------+
```

after

```
+--------------------+------------+------+----------+---------+---------+-------+---------+
|      filename      |  function  | line | time acc |   sd    |   cv    | calls | threads |
+--------------------+------------+------+----------+---------+---------+-------+---------+
| QueryMetricLog.cpp | startQuery |  163 |   1.34 s | 1.29 ms | 9.62412 | 10000 |    1000 |
+--------------------+------------+------+----------+---------+---------+-------+---------+
+----------+----------+----------+-----------+-----------+-----------+-----------+---------+----------+----------+
|    fastest[0-1]%    |                           center[1-99]%                            |  slowest[99-100]%   |
+----------+----------+----------+-----------+-----------+-----------+-----------+---------+----------+----------+
|   min    |   mean   |   min    |   mean    |    med    |  time a   |  time ae  |   max   |   mean   |   max    |
+----------+----------+----------+-----------+-----------+-----------+-----------+---------+----------+----------+
| 3.89 mcs | 4.91 mcs | 5.33 mcs | 19.38 mcs | 14.04 mcs | 173.85 ms | 173.85 ms | 1.05 ms | 11.55 ms | 25.61 ms |
+----------+----------+----------+-----------+-----------+-----------+-----------+---------+----------+----------+
```

##### finishQuery

before

```
+--------------------+-------------+------+----------+----------+---------+-------+---------+
|      filename      |  function   | line | time acc |    sd    |   cv    | calls | threads |
+--------------------+-------------+------+----------+----------+---------+-------+---------+
| QueryMetricLog.cpp | finishQuery |  155 |  97.51 s | 14.96 ms | 1.53452 | 10000 |    1000 |
+--------------------+-------------+------+----------+----------+---------+-------+---------+
+-----------+-----------+-----------+---------+------------+--------+-----------+----------+----------+-----------+
|     fastest[0-1]%     |                          center[1-99]%                           |   slowest[99-100]%   |
+-----------+-----------+-----------+---------+------------+--------+-----------+----------+----------+-----------+
|    min    |   mean    |    min    |  mean   |    med     | time a |  time ae  |   max    |   mean   |    max    |
+-----------+-----------+-----------+---------+------------+--------+-----------+----------+----------+-----------+
| 26.66 mcs | 30.80 mcs | 32.52 mcs | 9.11 ms | 744.27 mcs | 2.61 s | 896.82 ms | 63.45 ms | 81.95 ms | 133.44 ms |
+-----------+-----------+-----------+---------+------------+--------+-----------+----------+----------+-----------+
```

after

```
+--------------------+-------------+------+----------+---------+---------+-------+---------+
|      filename      |  function   | line | time acc |   sd    |   cv    | calls | threads |
+--------------------+-------------+------+----------+---------+---------+-------+---------+
| QueryMetricLog.cpp | finishQuery |  181 |  13.84 s | 4.20 ms | 3.03041 | 10000 |    1000 |
+--------------------+-------------+------+----------+---------+---------+-------+---------+
+-----------+-----------+-----------+---------+-----------+--------+----------+----------+----------+----------+
|     fastest[0-1]%     |                         center[1-99]%                          |  slowest[99-100]%   |
+-----------+-----------+-----------+---------+-----------+--------+----------+----------+----------+----------+
|    min    |   mean    |    min    |  mean   |    med    | time a | time ae  |   max    |   mean   |   max    |
+-----------+-----------+-----------+---------+-----------+--------+----------+----------+----------+----------+
| 26.92 mcs | 31.45 mcs | 33.38 mcs | 1.10 ms | 72.87 mcs | 1.86 s | 82.81 ms | 23.42 ms | 30.20 ms | 62.61 ms |
+-----------+-----------+-----------+---------+-----------+--------+----------+----------+----------+----------+
```

<!---
A technical comment, you are free to remove or leave it as it is when PR is created
The following categories are used in the next scripts, update them accordingly
utils/changelog/changelog.py
tests/ci/cancel_and_rerun_workflow_lambda/app.py
-->
### Changelog category (leave one):
- Performance Improvement


### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Improve the performance and accuracy of system.query_metric_log collection interval by reducing the critical region.

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

<!---
Directly edit documentation source files in the "docs" folder with the same pull-request as code changes

or

Add a user-readable short description of the changes that should be added to docs.clickhouse.com below.

At a minimum, the following information should be added (but add more as needed).
- Motivation: Why is this function, table engine, etc. useful to ClickHouse users?

- Parameters: If the feature being added takes arguments, options or is influenced by settings, please list them below with a brief explanation.

- Example use: A query or command.
-->


> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ] <!---ci_set_required--> Allow: All Required Checks
- [ ] <!---ci_include_stateless--> Allow: Stateless tests
- [ ] <!---ci_include_stateful--> Allow: Stateful tests
- [ ] <!---ci_include_integration--> Allow: Integration Tests
- [ ] <!---ci_include_performance--> Allow: Performance tests
- [ ] <!---ci_set_builds--> Allow: All Builds
- [ ] <!---batch_0_1--> Allow: batch 1, 2 for multi-batch jobs
- [ ] <!---batch_2_3--> Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ] <!---ci_exclude_style--> Exclude: Style check
- [ ] <!---ci_exclude_fast--> Exclude: Fast test
- [ ] <!---ci_exclude_asan--> Exclude: All with ASAN
- [ ] <!---ci_exclude_tsan|msan|ubsan|coverage--> Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ] <!---ci_exclude_aarch64|release|debug--> Exclude: All with aarch64, release, debug
---
- [ ] <!---ci_include_fuzzer--> Run only fuzzers related jobs (libFuzzer fuzzers, AST fuzzers, etc.)
- [ ] <!---ci_exclude_ast--> Exclude: AST fuzzers
---
- [ ] <!---do_not_test--> Do not test
- [ ] <!---woolen_wolfdog--> Woolen Wolfdog
- [ ] <!---upload_all--> Upload binaries for special builds
- [ ] <!---no_merge_commit--> Disable merge-commit
- [ ] <!---no_ci_cache--> Disable CI cache

pamarcos committed:
Reduce the general critical section for query_metric_log

- Use a separate mutex for each query to reduce the contention
period for queries_mutex.

- Refactor to use std::mutex instead of std::recursive_mutex
for queries_mutex.

- In case we're running late to schedule the next task, schedule
it immediately.

- Fix LockGuard because unlocking twice is undefined behavior.

robot-ch-test-poll4 commented:
<!-- automatic status comment for PR #71473 from ClickHouse/ClickHouse:improve-system-query-metric-log -->
*This is an automated comment for commit 6f624af5a8affce7c82c76b474288bc89ee64184 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/71473/6f624af5a8affce7c82c76b474288bc89ee64184/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Docs check</td><td>Builds and tests the documentation</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/bdd18b35b29f6078b620c65a4f4d59853cfe5371/docs_check.html">✅ success</a></td></tr>
<tbody>
</table>
</details>

pamarcos committed:
Small refactor to ease debugging when something happens on the CI

pamarcos committed:
Log when the query finishes for system.query_metric_log ASAP

There are logs where we can see that after the query finishes,
executeQuery takes up to 2 seconds to call finishQuery.

pamarcos committed:
Add a new setting query_metric_log_debug to avoid the noise

pamarcos committed:
Merge branch 'master' into improve-system-query-metric-log

pamarcos committed:
Revert "Add a new setting query_metric_log_debug to avoid the noise"

This reverts commit 955f537bd5ef2f4a29717ac4999ce2af47b4c039.

pamarcos committed:
Demote log from warning to debug to avoid failing the test

nickitat commented on src/Interpreters/QueryMetricLog.cpp:
```diff
@@ -139,51 +199,66 @@ void QueryMetricLog::finishQuery(const String & query_id, TimePoint finish_time,
     }
 
-    query_status.last_collect_time = query_info_time;
+    /// Leave some margin because task->scheduleAfter takes a value in milliseconds.
```
i don't think it will work in real world, actual wake up time might be postponed by a lot because of high load or just scheduling

pamarcos replied:
>actual wake up time might be postponed by a lot because of high load or just scheduling

I completely agree, but take into account that this checks exactly the opposite: that the task is never scheduled **before**  it's supposed to. Which is something that should never happen according to [this](https://github.com/ClickHouse/ClickHouse/blob/dc92242c39504bd21f756939854ae7d7ffc41252/src/Core/BackgroundSchedulePool.cpp#L340).

I did see some unexpected behavior, so I decided to add this just to check. I can add more margin just in case or even remove it if you deem so, but I think it `BackgroundSchedulePool` is doing its job, it should never get here.

nickitat replied:
> that the task is never scheduled before it's supposed to

sorry, i misunderstood it

pamarcos commented on src/Interpreters/QueryMetricLog.h:
```diff
@@ -36,31 +43,49 @@ struct QueryMetricLogElement
-    std::chrono::system_clock::time_point next_collect_time;
-    std::vector<ProfileEvents::Count> last_profile_events = std::vector<ProfileEvents::Count>(ProfileEvents::end());
-    BackgroundSchedulePool::TaskHolder task;
+    std::chrono::system_clock::time_point last_collect_time TSA_GUARDED_BY(getMutex());
```
I don't think we can annotate that for any modification of an object that a mutex within that same object needs to be locked. At least, I don't know how to do it.

I could wrap all elements within another struct and annotate only one instead, apart of the `scheduleNext` and `createLogMetricElement` methods.

pamarcos committed:
Refactor to avoid too much TSA_GUARD clutter

Remove LOGICAL_ERROR check that should never happen.
It didn't happen on the CI so far, and seems unnecessary.

pamarcos committed:
Merge branch 'master' into improve-system-query-metric-log

pamarcos committed:
Remove unnecessary extern

pamarcos committed:
Merge branch 'master' into improve-system-query-metric-log
